File: gateway-go/pkg/protocol/session.go
// Package protocol — session wire types.
//
// These types mirror the Protobuf definitions in proto/session.proto
// and the TypeScript types in src/protocol/generated/session.ts.
// The internal/session package uses its own Go-idiomatic types;
// these bridge the gap for RPC wire-format serialization.
package protocol

// SessionRunStatus mirrors proto/session.proto SessionRunStatus.
type SessionRunStatus string

const (
	SessionStatusRunning SessionRunStatus = "running"
	SessionStatusDone    SessionRunStatus = "done"
	SessionStatusFailed  SessionRunStatus = "failed"
	SessionStatusKilled  SessionRunStatus = "killed"
	SessionStatusTimeout SessionRunStatus = "timeout"
)

// SessionKind mirrors proto/session.proto SessionKind.
type SessionKind string

const (
	SessionKindDirect   SessionKind = "direct"
	SessionKindGroup    SessionKind = "group"
	SessionKindGlobal   SessionKind = "global"
	SessionKindUnknown  SessionKind = "unknown"
	SessionKindCron     SessionKind = "cron"
	SessionKindSubagent SessionKind = "subagent"
)

// ParseSessionKind converts a string to SessionKind, defaulting to direct.
func ParseSessionKind(s string) SessionKind {
	switch s {
	case "group":
		return SessionKindGroup
	case "global":
		return SessionKindGlobal
	case "unknown":
		return SessionKindUnknown
	case "cron":
		return SessionKindCron
	case "subagent":
		return SessionKindSubagent
	default:
		return SessionKindDirect
	}
}

// SessionLifecyclePhase mirrors proto/session.proto SessionLifecyclePhase.
type SessionLifecyclePhase string

const (
	SessionPhaseStart SessionLifecyclePhase = "start"
	SessionPhaseEnd   SessionLifecyclePhase = "end"
	SessionPhaseError SessionLifecyclePhase = "error"
)

// SessionTransition represents a state transition event.
// Mirrors proto/session.proto SessionTransition.
type SessionTransition struct {
	Key         string           `json:"key"`
	FromStatus  SessionRunStatus `json:"fromStatus"`
	ToStatus    SessionRunStatus `json:"toStatus"`
	TimestampMs int64            `json:"timestampMs"`
	Reason      string           `json:"reason,omitempty"`
	StopReason  string           `json:"stopReason,omitempty"`
	Aborted     *bool            `json:"aborted,omitempty"`
}

// SessionLifecycleEvent represents a session lifecycle event.
// Mirrors proto/session.proto SessionLifecycleEvent.
type SessionLifecycleEvent struct {
	Key         string                `json:"key"`
	Phase       SessionLifecyclePhase `json:"phase"`
	TimestampMs int64                 `json:"timestampMs"`
	StopReason  string                `json:"stopReason,omitempty"`
	Aborted     *bool                 `json:"aborted,omitempty"`
}

// SessionPreviewItem is a single item in a session preview.
// Mirrors proto/session.proto SessionPreviewItem.
type SessionPreviewItem struct {
	Role string `json:"role"` // "user" | "assistant" | "tool" | "system" | "other"
	Text string `json:"text"`
}

// SessionsPreviewEntry is a preview of a session.
// Mirrors proto/session.proto SessionsPreviewEntry.
type SessionsPreviewEntry struct {
	Key    string               `json:"key"`
	Status string               `json:"status"` // "ok" | "empty" | "missing" | "error"
	Items  []SessionPreviewItem `json:"items,omitempty"`
}

File: core-rs/core/src/protocol/session.rs
//! Session wire types.
//!
//! Mirrors `gateway-go/pkg/protocol/session.go`.

use serde::{Deserialize, Serialize};

/// Session run status.
/// Mirrors Go `SessionRunStatus`.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub enum SessionRunStatus {
    #[serde(rename = "running")]
    Running,
    #[serde(rename = "done")]
    Done,
    #[serde(rename = "failed")]
    Failed,
    #[serde(rename = "killed")]
    Killed,
    #[serde(rename = "timeout")]
    Timeout,
}

/// Session kind.
/// Mirrors Go `SessionKind`.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub enum SessionKind {
    #[serde(rename = "direct")]
    Direct,
    #[serde(rename = "group")]
    Group,
    #[serde(rename = "global")]
    Global,
    #[serde(rename = "unknown")]
    Unknown,
}

/// Parse a string into [`SessionKind`], defaulting to `Direct`.
///
/// Mirrors Go `ParseSessionKind`.
pub fn parse_session_kind(s: &str) -> SessionKind {
    match s {
        "group" => SessionKind::Group,
        "global" => SessionKind::Global,
        "unknown" => SessionKind::Unknown,
        _ => SessionKind::Direct,
    }
}

/// Session lifecycle phase.
/// Mirrors Go `SessionLifecyclePhase`.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub enum SessionLifecyclePhase {
    #[serde(rename = "start")]
    Start,
    #[serde(rename = "end")]
    End,
    #[serde(rename = "error")]
    Error,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_session_kind() {
        assert_eq!(parse_session_kind("group"), SessionKind::Group);
        assert_eq!(parse_session_kind("global"), SessionKind::Global);
        assert_eq!(parse_session_kind("unknown"), SessionKind::Unknown);
        assert_eq!(parse_session_kind("direct"), SessionKind::Direct);
        assert_eq!(parse_session_kind(""), SessionKind::Direct);
        assert_eq!(parse_session_kind("bogus"), SessionKind::Direct);
    }

    #[test]
    fn test_session_run_status_json() {
        assert_eq!(
            serde_json::to_string(&SessionRunStatus::Running).expect("serialize"),
            "\"running\""
        );
        assert_eq!(
            serde_json::to_string(&SessionRunStatus::Timeout).expect("serialize"),
            "\"timeout\""
        );
    }

    #[test]
    fn test_session_kind_roundtrip() {
        let kind = SessionKind::Group;
        let json = serde_json::to_string(&kind).expect("serialize");
        let parsed: SessionKind = serde_json::from_str(&json).expect("deserialize");
        assert_eq!(parsed, SessionKind::Group);
    }

    #[test]
    fn test_lifecycle_phase_json() {
        assert_eq!(
            serde_json::to_string(&SessionLifecyclePhase::Start).expect("serialize"),
            "\"start\""
        );
        assert_eq!(
            serde_json::to_string(&SessionLifecyclePhase::Error).expect("serialize"),
            "\"error\""
        );
    }
}
